File: TriePrefix.py
# TC: O(L) all operations
# SC: O(L) L-length of the word 

# We define each node with isEnd flag to indicate end of word and children array of 26 chars 
class TrieNode:
    def __init__(self):
        self.isEnd = False
        self.children = [None] * 26

class Trie:
    # new node creation
    def __init__(self):
        self.root = TrieNode()
        
    def insert(self, word: str) -> None:
        # start from root
        curr = self.root
        # check for each char of word if it exits already in the sequence, start creating sequence of nodes when a char doesnt exist
        for i in range(len(word)):
            ch = word[i]
            if curr.children[ord(ch)-ord('a')] == None:
                curr.children[ord(ch)-ord('a')] = TrieNode()
            # move pointer to its child node as found/created in prev step 
            curr = curr.children[ord(ch)-ord('a')]
        # mark end of word after loop ends
        curr.isEnd = True
        
    def search(self, word: str) -> bool:
        curr = self.root
        for i in range(len(word)):
            ch = word[i]
            if curr.children[ord(ch)-ord('a')] == None: return False
            curr = curr.children[ord(ch)-ord('a')]
        # return isEnd. For word to exist, isEnd should be true at the last char node of the word
        return curr.isEnd

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for i in range(len(prefix)):
            ch = prefix[i]
            if curr.children[ord(ch)-ord('a')] == None: return False
            curr = curr.children[ord(ch)-ord('a')]
        # prefix, we need not worry for isEnd check, return True if above loop is not broken
        return True

File: trie-prefix.java
//Time Complexity :O(L)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
class Trie {
    public class Node{
        char c;
        boolean isWord;
        Node[] children;
        public Node(char c){
            this.c = c;
            this.children = new Node[26];
            this.isWord = false;
        }
    }
    Node root;
    public Trie() {
        root  = new Node('\0');
    }
    public void insert(String word) {
        Node curr = root;
        for(int i =0;i<word.length();i++){
            char cc = word.charAt(i);
            if(curr.children[cc-'a'] == null) curr.children[cc-'a'] = new Node(cc);
            curr = curr.children[cc-'a'];
        }
        curr.isWord = true;
    }

    public boolean search(String word) {
        Node curr = root;
        for(int i = 0;i<word.length();i++){
            char cc = word.charAt(i);
            if(curr.children[cc-'a'] == null) return false;
            curr = curr.children[cc-'a'];
        }
        return curr.isWord == true?true:false;
    }

    public boolean startsWith(String word) {
        Node curr = root;
        for(int i = 0;i<word.length();i++){
            char cc = word.charAt(i);
            if(curr.children[cc-'a'] == null) return false;
            curr = curr.children[cc-'a'];
        }
        return true;
    }
}